collector/diskstats_darwin.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2017 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !nodiskstats
+
+package collector
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/lufia/iostat"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	diskSubsystem = "disk"
+)
+
+type typedDescFunc struct {
+	typedDesc
+	value func(stat *iostat.DriveStats) float64
+}
+
+type diskstatsCollector struct {
+	descs []typedDescFunc
+}
+
+func init() {
+	Factories["diskstats"] = NewDiskstatsCollector
+}
+
+// NewDiskstatsCollector returns a new Collector exposing disk device stats.
+func NewDiskstatsCollector() (Collector, error) {
+	var diskLabelNames = []string{"device"}
+
+	return &diskstatsCollector{
+		descs: []typedDescFunc{
+			{
+				typedDesc: typedDesc{
+					desc: prometheus.NewDesc(
+						prometheus.BuildFQName(Namespace, diskSubsystem, "reads_completed"),
+						"The total number of reads completed successfully.",
+						diskLabelNames,
+						nil,
+					),
+					valueType: prometheus.CounterValue,
+				},
+				value: func(stat *iostat.DriveStats) float64 {
+					return float64(stat.NumRead)
+				},
+			},
+			{
+				typedDesc: typedDesc{
+					desc: prometheus.NewDesc(
+						prometheus.BuildFQName(Namespace, diskSubsystem, "sectors_read"),
+						"The total number of sectors read successfully.",
+						diskLabelNames,
+						nil,
+					),
+					valueType: prometheus.CounterValue,
+				},
+				value: func(stat *iostat.DriveStats) float64 {
+					return float64(stat.NumRead) / float64(stat.BlockSize)
+				},
+			},
+			{
+				typedDesc: typedDesc{
+					desc: prometheus.NewDesc(
+						prometheus.BuildFQName(Namespace, diskSubsystem, "read_time_ms"),
```

Comment:
Agreed with @grobie, all new metrics should follow the best practices (unless maybe all the exporter is doing is surfacing existing metrics name).